applications: nrf_desktop: Adding a new sensor (#2333)
nrfconnect/sdk-nrf: applications/nrf_desktop/README.rst

   Some options have dependencies that might not be needed when these options are disabled.
   For example, when the LEDs module is disabled, the PWM driver is not needed.

Adding a new motion sensor
==========================

The nRF52 Desktop application comes with a motion module able to read data from a motion sensor.
A communication with the sensor is done through a sensor driver API.
In the scope of NCS two motion sensor drivers are available, for PMW3360 and PAW3212.
It is likely that another sensor will be more suitable for a new board.
In this section it is described how to add a new motion sensor into the project.

The first step is to create a new motion sensor driver, that will provide code communicating with the sensor.
Driver shall provide the sensor API (see :ref:`sensor_api`), that will be used by the application.
The two existing drivers can be used as an example.
For motion module to correctly work with a driver, it must support a trigger on a new data.

> When data is ready the motion module will call data fetch and data get on two sensor channels ``SENSOR_CHAN_POS_DX`` and ``SENSOR_CHAN_POS_DX``.

Reviewer: DX and DY

Author: fixed

The driver must support these two channels with its data fetch and data get functions.

The motion sensor configuration should be done through the DTS.
To allow such configuration the DTS binding must be defined.
Please refer to :file:`dts/bindings/sensor` for binding examples.
When binding is ready a DTS file describing the board can refer to it while depicting the sensor configuration.
Communication with both PMW3360 and PAW3212 is done through the SPI.
In case of both gaming and desktop mouse, that are part of this reference design, both sensors lie as a node of SPI bus (look for ``pmw3360`` in :file:`boards/arm/nrf52840gmouse_nrf52840/nrf52840gmouse_nrf52840.dts`).
In case a different kind of bus is used for the new sensor, the DTS layout will be different.

.. note::

>     To achive the full speed, the data must be propagated through the application and reach the Bluetooth LE a few houndred milliseconds before the subsequent connection event.

Reviewer: microseconds :)

Author: fixed

    In case the lowest latency through the LLPM is desired (1-ms interval), it is important that sensor data readout takes no more then 250 us.
    The bus and the sensor configuration must assure that communication speed is fast enough.

When a raw support for the new sensor was added (both the driver and the sensor configuration are in place), the new sensor must now be included in the application.
The nRF52 Desktop application selects a sensor using configuration options defined in :file:`src/hw_interface/Kconfig.motion`.
The new sensor should be added as a new choice.
The last step is to modify :file:`configuration/common/motion_sensor.h` to translate sensor specific attributes to a generic abstraction.

> If attribute is not supported by the sensor it does not have to be defined.

Reviewer: Should be set to `-ENOTSUP`

Author: Fixed

   Some options have dependencies that might not be needed when these options are disabled.
   For example, when the LEDs module is disabled, the PWM driver is not needed.

> Adding a new motion sensor

Author: Done

> Adding a new motion sensor

Author: > The first paragraphs don't tell me why I would want to add a new sensor.

You had it in the first paragraph and even made a comment :)
> It is likely that another sensor will be more suitable for a new board.

> Adding a new motion sensor

Reviewer 2: > It is likely that another sensor will be more suitable for a new board.

This doesn't serve this purpose, because as written earlier, it is a supposition, not a statement of fact. This `likely` poses questions like "why will it be more suitable? what is wrong with the NCS sensors? what another sensor should be added?". None of these questions is answered here.

==========================

> The nRF52 Desktop application comes with a motion module able to read data from a motion sensor.

Author: Added

> The nRF52 Desktop application comes with a motion module able to read data from a motion sensor.

Author: Fixed

A communication with the sensor is done through a sensor driver API.

> In the scope of NCS two motion sensor drivers are available, for PMW3360 and PAW3212.

Author: We don't have documentation for these sensor drivers.

> It is likely that another sensor will be more suitable for a new board.

Author: Client will most probably use different sensor. It's not a doubt but the fact that other sensor that we don't have support for will most likely be used. But if it will be we cannot be 100% sure.
I will try to write it differently/

> It is likely that another sensor will be more suitable for a new board.

Author: I removed this sentence altogether.

> In this section it is described how to add a new motion sensor into the project.

Author: I think this is needed to start the description. If I throw it away, what comes next will sound strange.

> In this section it is described how to add a new motion sensor into the project.

Author: Used the suggestion below.

> The first step is to create a new motion sensor driver, that will provide code communicating with the sensor.

Author: Done

> Driver shall provide the sensor API (see :ref:`sensor_api`), that will be used by the application.

Author: Driver can, mar or should use this API but if it will not use it the whole thing will not work. So it either shall or must use it.

The two existing drivers can be used as an example.
For motion module to correctly work with a driver, it must support a trigger on a new data.

> When data is ready the motion module will call data fetch and data get on two sensor channels ``SENSOR_CHAN_POS_DX`` and ``SENSOR_CHAN_POS_DY``.

Reviewer 2: `When data is ready` - can we be more specific here?

Author: I have changed it to "When motion data is ready".

> The driver must support these two channels with its data fetch and data get functions.

Reviewer 2: How to ensure this?

Author: This is not a problem. The application will indicate an error if these are not supported.

The motion sensor configuration should be done through the DTS.
To allow such configuration the DTS binding must be defined.
Please refer to :file:`dts/bindings/sensor` for binding examples.

> When binding is ready a DTS file describing the board can refer to it while depicting the sensor configuration.

Reviewer 2: `is ready` - change to `is defined`, so that the link to line 975 is clear?

Author: done

> When binding is ready a DTS file describing the board can refer to it while depicting the sensor configuration.

Reviewer 2: `depicting` -- what do you mean here exactly? A figure is drawn?

Author: https://www.thefreedictionary.com/depicting
> 2. To represent in words; describe:

I meant the second meaning of the word. In this case describing how the sensor HW is configured.

Reviewer 2: `describe` is the word that should be used here. `depict` belongs to different register and context.

> Communication with both PMW3360 and PAW3212 is done through the SPI.

Reviewer 2: Is there anything to do here? The text suggests all in lines 978-980 happens automatically. Knowing your style a bit already, you want to ask the user to configure something here, am I correct?

Author: Yep. I can tell what we did for our sensors but it is obvious when one read the config file. My goal was to say - fine line x in file y and read it. Since the dts language and layout is more or less self explaining one should get it without any problem.
Configuring another SPI sensor will be very easy from this point. As for other sensors... this depends what these use. I added ref to DTS into - this should help.

Author: I have expanded this section

> In case of both gaming and desktop mouse, that are part of this reference design, both sensors lie as a node of SPI bus (look for ``pmw3360`` in :file:`boards/arm/nrf52840gmouse_nrf52840/nrf52840gmouse_nrf52840.dts`).

Reviewer 2: Can we link to the desktop mouse and gaming mouse sections? Either architecture or user interface?

Author: Architecture does not go into detail how sensor is connected. That will not give us much as there is simply a line between our module, a sensor driver and and SPI driver. Still we can make a ref.

Reviewer 2: OK

In case a different kind of bus is used for the new sensor, the DTS layout will be different.

.. note::
    To achive the full speed, the data must be propagated through the application and reach the Bluetooth LE a few houndred microseconds before the subsequent connection event.
    In case the lowest latency through the LLPM is desired (1-ms interval), it is important that sensor data readout takes no more then 250 us.

>     The bus and the sensor configuration must assure that communication speed is fast enough.

Reviewer 2: `assure` - `ensure` or `make sure`

Author: Fixed

> When a raw support for the new sensor was added (both the driver and the sensor configuration are in place), the new sensor must now be included in the application.

Reviewer 2: See comment to line 978. Also, do not use past tense here.

Author: Done

> The nRF52 Desktop application selects a sensor using configuration options defined in :file:`src/hw_interface/Kconfig.motion`.

Reviewer 2: nRF52?

Author: Fixed

The new sensor should be added as a new choice.
The last step is to modify :file:`configuration/common/motion_sensor.h` to translate sensor specific attributes to a generic abstraction.
If attribute is not supported by the sensor it does not have to be defined.
In such case it should be set to ``-ENOTSUP``.

> 

Reviewer 2: What is the result of this configuration process? How does the sensor behave after completing all steps? Please add a short summary.

Author: Added one more step

   Some options have dependencies that might not be needed when these options are disabled.
   For example, when the LEDs module is disabled, the PWM driver is not needed.

Adding a new motion sensor
==========================

The nRF Desktop application comes with a :ref:`nrf_desktop_motion` able to read data from a motion sensor.
In the scope of NCS two motion sensor drivers are available, for PMW3360 and PAW3212.
Complete the following steps to add a new motion sensor into the project.

1. Add a new sensor driver

   The first step is to create a new motion sensor driver, that will provide code communicating with the sensor.
   A communication between the application and the sensor is done through a sensor driver API (see :ref:`sensor_api`).
   For motion module to work correctly, the driver must support a trigger (see ``sensor_trigger_set``) on a new data (see ``SENSOR_TRIG_DATA_READY`` trigger type).

>    When motion data is ready, the driver calls a register callback.

Reviewer: Only the first sample after inactivity is fetched this way, then samples are fetched by the thread without using the trigger.

>    When motion data is ready, the driver calls a register callback.

Author: Thanks, I have almost forgot ( ͡° ͜ʖ ͡°)

>    When motion data is ready, the driver calls a register callback.

Author: Seriously though I don't want to repeat the pipeline process for I think third time. The focus here is on how upper layer interacts with the driver.

>    When motion data is ready, the driver calls a register callback.

Reviewer: I just meant that this suggests the reader that sensor reads every sample using the callback (this could be a bit misleading). But we can leave it as it is.

   .. code-block:: DTS

      &spi1 {
      	compatible = "nordic,nrf-spim";
      	status = "okay";
      	sck-pin = <16>;
      	mosi-pin = <17>;
      	miso-pin = <15>;
      	cs-gpios = <&gpio0 13 0>;

      	pmw3360@0 {
      		compatible = "pixart,pmw3360";
      		reg = <0>;
      		irq-gpios = <&gpio0 21 0>;
      		spi-max-frequency = <2000000>;
      		label = "PMW3360";
      	};
      };

   Communication with PMW3360 is done through the SPI, making the sensor a subnode of SPI bus node.
   SPI pins are defined as part of the bus configuration, as these are common among all devices connected to this bus.
   In our case the PMW3360 is the only device on this bus and so there is only one chip select pin specified.
   When the sensor's node is referred you can read `@0` in `pmw3360@0`.
   The part `@0` in case of SPI devices refers to the position of the chip select pin in the `cs-gpios` array for a corresponding device.
   Note the string `compatible = "pixart,pmw3360"` in the subnode configuration.
   This indicates which DTS binding will the node use and it should match with the DTS binding created earlier for the sensor.

>    The options `req`, `label` and `spi-max-frequency` are inherited from the `spi-device` binding and are common to all SPI devices:

Reviewer: reg

>      * `req` is the slave ID number the device has on a bus.

Reviewer: reg

     * `label` is used to generate a name of the device (e.g. will be added to generated macros).
     * `spi-max-frequency` sets the bus clock frequency.

       .. note::

>            To achive the full speed, the data must be propagated through the application and reach the Bluetooth LE a few houndred microseconds before the subsequent connection event.

Reviewer: hundred

   The remaining option `irq-gpios` is specific to `pixart,pmw3360` binding.
   It refers to the PIN to which the motion sensor IRQ line is connected.

   In case a different kind of bus is used for the new sensor, the DTS layout will be different.

#. Sensor support in the nRF Desktop application

   After the new sensor is supported by |NCS| it can be included in the application.
   The nRF Desktop application selects a sensor using configuration options defined in :file:`src/hw_interface/Kconfig.motion`.
   The new sensor should be added as a new choice.
   The last step is to modify :file:`configuration/common/motion_sensor.h` to translate sensor specific attributes to a generic abstraction.
   If attribute is not supported by the sensor it does not have to be defined.
   In such case it should be set to ``-ENOTSUP``.

#. Selecting the new sensor

   After all the steps defined above are done, the new sensor can be used by the application.
   The application configuration files for your board (see :ref:`nrf_desktop_board_configuration`) can be modified to enable it.

   To use the new sensor:

>      * Enabled all dependencies required by the driver (e.g. bus driver).

Reviewer: Enable

4. Include sensor in the application
------------------------------------

Once the new sensor is supported by |NCS| and board configuration is updated, you can include it in the nRF Desktop application.

The nRF Desktop application selects a sensor using the configuration options defined in :file:`src/hw_interface/Kconfig.motion`.
Add the new sensor as a new choice option.

> The nRF Desktop application :ref:`nrf_desktop_motion` access several sensor attributes.

Reviewer 2: ```suggestion
The :ref:`nrf_desktop_motion` of the nRF Desktop application has access to several sensor attributes.
```

These attributes are used to modify the sensor behavior in runtime.

> Since the names on the attributes differ between sensors the :ref:`nrf_desktop_motion` uses generic abstraction of them.

Reviewer 2: ```suggestion
Since the names of the attributes differ for each sensor, the :ref:`nrf_desktop_motion` uses a generic abstraction of them.
```

> By modifying :file:`configuration/common/motion_sensor.h` the new sensor-specific attributes can be translated to a generic abstraction.

Reviewer 2: ```suggestion
You can translate the new sensor-specific attributes to a generic abstraction by modifying :file:`configuration/common/motion_sensor.h` .
```
